factcast-store-pgsql/src/test/java/org/factcast/store/pgsql/internal/listen/PgListenerTest.java

```diff
@@ -46,108 +49,154 @@
 
-    @Mock
+    @Mock(answer = Answers.RETURNS_DEEP_STUBS)
     PgConnection conn;
 
     @Mock
     PreparedStatement ps;
 
-    Predicate<Connection> tester = c -> true;
+    PGNotification[] someNotification = new PGNotification[] { //
+            new Notification(PgConstants.CHANNEL_NAME, 1) };
 
     @Captor
-    ArgumentCaptor<PgListener> captor;
+    ArgumentCaptor<FactInsertionEvent> factCaptor;
 
     @Test
-    void testCheckFails() throws SQLException {
-
+    void testRegisterPostgresListeners() throws Exception {
         when(ds.get()).thenReturn(conn);
         when(conn.prepareStatement(anyString())).thenReturn(ps);
+        when(conn.getNotifications(anyInt())).thenReturn(null);
+        when(conn.prepareCall(anyString()).execute()).thenReturn(true);
 
-        tester = mock(Predicate.class);
-        when(tester.test(any())).thenReturn(false, false, true, false);
-        PgListener l = new PgListener(ds, bus, tester);
+        PgListener l = new PgListener(ds, bus);
         l.afterPropertiesSet();
-        verify(bus, times(3)).post(any(FactInsertionEvent.class));
-        verifyNoMoreInteractions(bus);
-
+        sleep(100);
         l.destroy();
-    }
 
-    @Test
-    void testListen() throws Exception {
-        when(ds.get()).thenReturn(conn);
-        when(conn.prepareStatement(anyString())).thenReturn(ps);
-
-        PgListener l = new PgListener(ds, bus, tester);
-        l.afterPropertiesSet();
-        sleep(100);
         verify(bus).post(any(FactInsertionEvent.class));
         verifyNoMoreInteractions(bus);
         verify(conn, atLeastOnce()).prepareStatement(PgConstants.LISTEN_SQL);
+        verify(conn, atLeastOnce()).prepareStatement(PgConstants.LISTEN_ROUNDTRIP_CHANNEL_SQL);
         verify(ps, atLeastOnce()).execute();
-
-        l.destroy();
     }
 
     @Test
     void testNotify() throws Exception {
-
         when(ds.get()).thenReturn(conn);
         when(conn.prepareStatement(anyString())).thenReturn(ps);
-
-        PgListener l = new PgListener(ds, bus, tester);
+        when(conn.prepareCall(anyString()).execute()).thenReturn(true);
         when(conn.getNotifications(anyInt())).thenReturn(new PGNotification[] { //
                 new Notification(PgConstants.CHANNEL_NAME, 1), //
                 new Notification(PgConstants.CHANNEL_NAME, 1), //
                 new Notification(PgConstants.CHANNEL_NAME, 1) },
                 new PGNotification[] { new Notification(PgConstants.CHANNEL_NAME, 2) }, //
-                new PGNotification[] { new Notification(PgConstants.CHANNEL_NAME, 3) }, null);
-        l.afterPropertiesSet();
-        sleep(400);
-        // 4 posts: one scheduled, 3 from notifications
-        verify(bus, times(4)).post(any(FactInsertionEvent.class));
+                new PGNotification[] { new Notification(PgConstants.CHANNEL_NAME, 3) }, null, null,
+                null);
 
+        PgListener l = new PgListener(ds, bus);
+        l.afterPropertiesSet();
+        sleep(500);
```

Comment:
this creates longrunning or flaky tests, lets see if we can avoid that

Reply:
please have a look again. most of this stuff is gone. however, there is one test left where I am not sure (see TODO)